### PR TITLE
feat: jupiter auto-slippage

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

This PR brings in auto-slippage for Jupiter, and coincidentally removes the notion of default slippage always existing for swappers, since it does not for Jupiter anymore, and eventually won't for other swappers either.

```
TODO:

- [x] make it work
- [x] ensure manual slippage is still happy
- [x] cleanup Solana auto-slippage
```

## Issue (if applicable)

- closes https://github.com/shapeshift/web/issues/8246
- contributes to https://github.com/shapeshift/web/issues/8193 i.e this is the first swapper with auto-slippage support

## Risk

> High Risk PRs Require 2 approvals

Low, isolated to Jupiter, though technically, code paths that touch other swappers have been touched too 
> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing

- Without inputting any custom slippage, Jupiter slippage is dynamic, i.e the slippage shown (and used for the *quote* and its execution) are the ones gotten from the rate. This can be verified by
  - ensuring that `autoSlippage=true` is passed in the jup.ag `/quote` request when autoslippage is used in app
  - at final quote time, the `quoteResponse` property of the `/swap-instructions` request includes the same slippage bps that was previously gotten
  - from a UX perspective, the slippage displayed in slippage popover and in swapper (input/confirm) is the one from the rate
- With custom slippage inputted in, the slippage we pass to Jupiter honours the choice of the user. This can be verified by
  - ensuring `slippageBps` is used in place of `autoSlippage=true` for the `/quote` call
  - at final quote time, the `quoteResponse` property of the `/swap-instructions` request includes the same slippage bps that was previously inputted
  - from a UX perspective, the slippage displayed in slippage popover and in swapper (input/confirm) is the one the user chose

### Engineering

- ☝🏽 

### Operations

- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

- ☝🏽 

## Screenshots (if applicable)

https://jam.dev/c/8b726809-3883-4019-9279-e3f9a54b35c2

### CODE REVIEW

File: packages/swapper/src/utils.ts [hunk 1]
```diff
@@ -286,7 +286,7 @@ export const checkEvmSwapStatus = async ({
   }
 }
 
-export const getRate = ({
+export const getInputOutputRate = ({
```
Comment (gomesalexandre) [inline packages/swapper/src/utils.ts:289]: Renamed to a more declarative version so we don't confuse this with what we call `rates` e.g a `rate quote`

File: packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeRate/getPortalsTradeRate.tsx [hunk 2]
```diff
@@ -113,7 +113,7 @@ export async function getPortalsTradeRate(
     })
     // Use the quote estimate endpoint to get a quote without a wallet
 
-    const rate = getRate({
+    const rate = getInputOutputRate({
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeRate/getPortalsTradeRate.tsx:116]:
```suggestion
    const inputOutputRate = getInputOutputRate({
```

File: packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts [hunk 3]
```diff
@@ -135,7 +135,7 @@ export async function getPortalsTradeQuote(
 
     if (!tx) throw new Error('Portals Tx simulation failed upstream')
 
-    const rate = getRate({
+    const rate = getInputOutputRate({
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeQuote/getPortalsTradeQuote.ts:138]:
```suggestion
    const inputOutputRate = getInputOutputRate({
```

File: packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeRate.ts [hunk 4]
```diff
@@ -141,42 +141,41 @@ export const getTradeRate = async (
...
 
-  const getQuoteRate = (sellAmountCryptoBaseUnit: string, buyAmountCryptoBaseUnit: string) => {
-    return getRate({
-      sellAmountCryptoBaseUnit,
-      buyAmountCryptoBaseUnit,
-      sellAsset,
-      buyAsset,
-    })
-  }
-
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeRate.ts:152]: Not sure why we curried this, removed to keep things clean

File: packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeRate.ts [hunk 5]
```diff
@@ -96,7 +96,7 @@ export const getTradeRate = async (
     )
   }
 
-  const { data: quoteResponse } = maybePriceResponse.unwrap()
+  const { data: priceResponse } = maybePriceResponse.unwrap()
```
Comment (gomesalexandre) [inline packages/swapper/src/swappers/JupiterSwapper/swapperApi/getTradeRate.ts:99]: i.e this isn't a quote

File: packages/swapper/src/constants.ts [hunk 6]
```diff
@@ -105,10 +105,9 @@ const DEFAULT_LIFI_SLIPPAGE_DECIMAL_PERCENTAGE = '0.005' // .5%
 const DEFAULT_THOR_SLIPPAGE_DECIMAL_PERCENTAGE = '0.01' // 1%
 const DEFAULT_ARBITRUM_BRIDGE_SLIPPAGE_DECIMAL_PERCENTAGE = '0' // no slippage for Arbitrum Bridge, so no slippage tolerance
 const DEFAULT_CHAINFLIP_SLIPPAGE_DECIMAL_PERCENTAGE = '0.02' // 2%
-const DEFAULT_JUPITER_SLIPPAGE_DECIMAL_PERCENTAGE = '0.01' // 1%
 
 export const getDefaultSlippageDecimalPercentageForSwapper = (
-  swapperName?: SwapperName,
+  swapperName: SwapperName | undefined,
```
Comment (gomesalexandre) [inline packages/swapper/src/constants.ts:110]: We never call this without args, we really meant Maybe here

File: packages/swapper/src/constants.ts [hunk 7]
```diff
@@ -128,8 +127,17 @@ export const getDefaultSlippageDecimalPercentageForSwapper = (
     case SwapperName.Chainflip:
       return DEFAULT_CHAINFLIP_SLIPPAGE_DECIMAL_PERCENTAGE
     case SwapperName.Jupiter:
-      return DEFAULT_JUPITER_SLIPPAGE_DECIMAL_PERCENTAGE
+      throw new Error('Default slippage not supported by Jupiter')
     default:
-      assertUnreachable(swapperName)
+      return assertUnreachable(swapperName)
```
Comment (gomesalexandre) [inline packages/swapper/src/constants.ts:132]: tsc wasn't very happy without actually returning (already the case in develop) though not actually throwing `tsc` compile errors

File: packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeRate/getPortalsTradeRate.tsx (hunk 2 above)
Comment (gomesalexandre) [reply packages/swapper/src/swappers/PortalsSwapper/getPortalsTradeRate/getPortalsTradeRate.tsx]: ce38a8d8f

File: src/components/MultiHopTrade/components/SharedTradeInput/SharedSlippagePopover.tsx [hunk 8]
```diff
@@ -38,7 +38,7 @@ const focusStyle = { '&[aria-invalid=true]': { borderColor: 'red.500' } }
 const faGear = <FaGear />
 
 type SharedSlippagePopoverProps = {
-  defaultSlippagePercentage: string
+  defaultSlippagePercentage: string | undefined
```
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/SharedTradeInput/SharedSlippagePopover.tsx:41]: Removes the assumption that a default exists

**Review by NeOMakinG — APPROVED** [review]
SOL => WUF
image

MANEKI => WUF
image

WUF => HAPPY
image

COOLCAT => PEPE
image

A few TX:

https://explorer.solana.com/tx/62kdzzHBFKWsabpmDLvM9864jKVrHRmQysv9Wq8WnkiRSLznN8GXKrKqY93prcxuCKGvnKy5qZNysNEfrvLK9Y46
image

https://explorer.solana.com/tx/52EdsVBbxcrpE6XP7of3exjYnSiHhDjkQ8FF1CRbe8SFfQtkjowthV8rFZGxwPVGBhsFCmuGeaXGTJPhp6TLSo7M

image
https://explorer.solana.com/tx/4hKr2RRGwAF16UGbci57EviA7E9VT2cTHuymtRZfY2sRzfGfMH8fBexdEPzSMF1p1rJqkkqyRqveXMN7nfB9jaEY

Looks very good!

**Pull request merged**